File: case2/c2_simulate_openloop.py
#!/usr/bin/env python3
# Technique: Case 2 - Minimum Jerk QP Trajectory Planning + 6-Parameter Pseudo-Flatness Reconstruction (m11 != m22)

import os
import time
import json
import numpy as np
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

from c2_usv_params import (
    m11_real, m22_real, m33_real, Xu_real, Yv_real, Nr_real,
    dP, DT_SIM, T_MAX, T_MIN
)
from c2_min_jerk_qp import MinJerkTrajectory2D
from c2_flatness_reconstruct import reconstruct_flatness_h2

plt.rcParams['font.sans-serif'] = 'DejaVu Sans'
plt.rcParams['axes.edgecolor'] = '#333333'
plt.rcParams['axes.linewidth'] = 1.2

def real_6param_rk4_step(state, Tu, Tr, dt):
    def deriv(s):
        _, _, p_i, u_i, v_i, r_i = s
        dx = u_i * np.cos(p_i) - v_i * np.sin(p_i)
        dy = u_i * np.sin(p_i) + v_i * np.cos(p_i)
        dpsi = r_i
        du = (Tu + m22_real * v_i * r_i - Xu_real * u_i) / m11_real
        dv = (-m11_real * u_i * r_i - Yv_real * v_i) / m22_real
        dr = (Tr + (m11_real - m22_real) * u_i * v_i - Nr_real * r_i) / m33_real
        return np.array([dx, dy, dpsi, du, dv, dr])

    k1 = deriv(state)
    k2 = deriv(state + 0.5 * dt * k1)
    k3 = deriv(state + 0.5 * dt * k2)
    k4 = deriv(state + dt * k3)

    return state + (dt / 6.0) * (k1 + 2 * k2 + 2 * k3 + k4)

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))

    waypoints = np.array([
        [0.0, 0.0],
        [8.0, 0.0],
        [14.0, 5.0],
        [14.0, 13.0],
        [20.0, 17.0],
        [28.0, 17.0],
        [32.0, 10.0],
        [26.0, 4.0],
        [20.0, 1.5]
    ])

    base_times = np.array([0.0, 7.0, 14.0, 20.0, 27.0, 33.0, 40.0, 48.0, 55.0])
    time_scale = 2.10
    times = base_times * time_scale

    v0_vec = (0.1, 0.0)
    dir_f = waypoints[8] - waypoints[7]
    dir_f_unit = dir_f / np.linalg.norm(dir_f)
    vf_vec = tuple(0.01 * dir_f_unit)

    t_start = time.perf_counter()
    planner = MinJerkTrajectory2D(waypoints, times, vel_start=v0_vec, vel_end=vf_vec)
    t_sim, pos, vel, acc, jerk = planner.sample(dt=DT_SIM)
    flat_data = reconstruct_flatness_h2(pos, vel, acc, jerk, t_sim)
    solve_time_ms = (time.perf_counter() - t_start) * 1000.0
    n_steps = len(t_sim)


    eta_ref = flat_data['eta']
    nu_ref = flat_data['nu']
    tau_ref = flat_data['tau_plan']
    tau_act_flat = flat_data['tau_act']
    T_plan = flat_data['T_plan']
    T_act_flat = flat_data['T_act']

    X_ref_all = np.column_stack([eta_ref, nu_ref])

    state_real = X_ref_all[0].copy()

    hist_state_real = [state_real.copy()]
    hist_tau_applied = []
    hist_T_applied = []

    for i in range(n_steps - 1):
        Tu_apply, Tr_apply = tau_act_flat[i, 0], tau_act_flat[i, 1]
        hist_tau_applied.append([Tu_apply, Tr_apply])
        hist_T_applied.append([T_act_flat[i, 0], T_act_flat[i, 1]])

        state_real = real_6param_rk4_step(state_real, Tu_apply, Tr_apply, DT_SIM)
        hist_state_real.append(state_real.copy())

    hist_state_real = np.array(hist_state_real)
    hist_tau_applied = np.array(hist_tau_applied)
    hist_T_applied = np.array(hist_T_applied)

    pos_err = np.hypot(hist_state_real[:, 0] - eta_ref[:, 0], hist_state_real[:, 1] - eta_ref[:, 1])
    rmse_pos = float(np.sqrt(np.mean(pos_err**2)))
    max_err_pos = float(np.max(pos_err))

    psi_err = np.arctan2(np.sin(hist_state_real[:, 2] - eta_ref[:, 2]), np.cos(hist_state_real[:, 2] - eta_ref[:, 2]))
    rmse_psi_rad = float(np.sqrt(np.mean(psi_err**2)))
    rmse_psi_deg = float(np.degrees(rmse_psi_rad))

    rmse_u = float(np.sqrt(np.mean((hist_state_real[:, 3] - nu_ref[:, 0])**2)))
    rmse_v = float(np.sqrt(np.mean((hist_state_real[:, 4] - nu_ref[:, 1])**2)))
    rmse_r = float(np.sqrt(np.mean((hist_state_real[:, 5] - nu_ref[:, 2])**2)))

    metrics = {
        "case": "Case 2",
        "solver_type": "QP (6-Param Pseudo-Flatness)",
        "trajectory_solver_time_ms": solve_time_ms,
        "tracking_error": {
            "rmse_position_m": rmse_pos,
            "max_position_error_m": max_err_pos,
            "rmse_heading_rad": rmse_psi_rad,
            "rmse_heading_deg": rmse_psi_deg,
            "rmse_surge_u_mps": rmse_u,
            "rmse_sway_v_mps": rmse_v,
            "rmse_yaw_rate_r_radps": rmse_r
        }
    }

    metrics_path = os.path.join(script_dir, 'openloop_metrics.json')
    with open(metrics_path, 'w') as f:
        json.dump(metrics, f, indent=4)

    print("\n==================================================")
    print("CASE 2 OPEN-LOOP TRAJECTORY METRICS:")
    print(f"  - Trajectory Solver Time (QP): {solve_time_ms:.4f} ms")
    print(f"  - Position RMSE:               {rmse_pos:.4f} m")
    print(f"  - Max Position Error:          {max_err_pos:.4f} m")
    print(f"  - Heading RMSE:                {rmse_psi_rad:.4f} rad ({rmse_psi_deg:.2f} deg)")
    print(f"  - Surge Velocity (u) RMSE:     {rmse_u:.4f} m/s")
    print(f"  - Sway Velocity (v) RMSE:      {rmse_v:.4f} m/s")
    print(f"  - Yaw Rate (r) RMSE:           {rmse_r:.4f} rad/s")
    print(f"Metrics saved to: {metrics_path}")
    print("==================================================\n")

    fig, axs = plt.subplots(6, 1, figsize=(10, 19), dpi=300)

    c_plan = '#2563EB'
    c_real = '#DC2626'
    c_green_plan = '#059669'
    c_green_real = '#047857'
    c_amber_plan = '#D97706'

    t_ctrl = t_sim[:-1]

    axs[0].plot(eta_ref[:, 0], eta_ref[:, 1], color=c_plan, lw=2.2, ls='--', label='Planned')
    axs[0].plot(hist_state_real[:, 0], hist_state_real[:, 1], color=c_real, lw=2.0, label='Real')
    axs[0].scatter(waypoints[:, 0], waypoints[:, 1], color='#111827', s=50, zorder=5, label='Waypoints')
    for idx_wp, (wx, wy) in enumerate(waypoints):
        axs[0].annotate(f'WP{idx_wp}', (wx, wy), textcoords="offset points", xytext=(5, 5), fontsize=8, fontweight='bold')
    axs[0].set_xlabel('X [m]', fontweight='bold')
    axs[0].set_ylabel('Y [m]', fontweight='bold')
    axs[0].set_title('2D Trajectory', fontsize=11, fontweight='bold', pad=6)
    axs[0].grid(True, ls=':', alpha=0.6)
    axs[0].axis('equal')
    axs[0].legend(loc='best', frameon=True, facecolor='white', fontsize=8.5)

    axs[1].plot(t_sim, nu_ref[:, 0], color=c_plan, ls='--', lw=1.8, label='Planned')
    axs[1].plot(t_sim, hist_state_real[:, 3], color=c_real, lw=1.8, label='Real')
    axs[1].set_xlabel('Time [s]', fontweight='bold')
    axs[1].set_ylabel('Surge $u$ [m/s]', fontweight='bold')
    axs[1].set_title('Surge Velocity', fontsize=11, fontweight='bold', pad=6)
    axs[1].grid(True, ls=':', alpha=0.6)
    axs[1].legend(loc='upper right', frameon=True, facecolor='white', fontsize=8.5)

    axs[2].plot(t_sim, nu_ref[:, 1], color=c_green_plan, ls='--', lw=1.8, label='Planned')
    axs[2].plot(t_sim, hist_state_real[:, 4], color=c_real, lw=1.8, label='Real')
    axs[2].set_xlabel('Time [s]', fontweight='bold')
    axs[2].set_ylabel('Sway $v$ [m/s]', fontweight='bold')
    axs[2].set_title('Sway Velocity', fontsize=11, fontweight='bold', pad=6)
    axs[2].grid(True, ls=':', alpha=0.6)
    axs[2].legend(loc='upper right', frameon=True, facecolor='white', fontsize=8.5)

    axs[3].plot(t_sim, nu_ref[:, 2], color=c_amber_plan, ls='--', lw=1.8, label='Planned')
    axs[3].plot(t_sim, hist_state_real[:, 5], color=c_real, lw=1.8, label='Real')
    axs[3].set_xlabel('Time [s]', fontweight='bold')
    axs[3].set_ylabel('Yaw Rate $r$ [rad/s]', fontweight='bold')
    axs[3].set_title('Yaw Rate', fontsize=11, fontweight='bold', pad=6)
    axs[3].grid(True, ls=':', alpha=0.6)
    axs[3].legend(loc='upper right', frameon=True, facecolor='white', fontsize=8.5)

    axs[4].plot(t_sim, tau_ref[:, 0], color=c_plan, ls='--', lw=1.8, label='$\\tau_u$ Planned')
    axs[4].plot(t_ctrl, hist_tau_applied[:, 0], color=c_real, lw=1.8, label='$\\tau_u$ Applied')
    axs[4].plot(t_sim, tau_ref[:, 1], color=c_amber_plan, ls='--', lw=1.8, label='$\\tau_r$ Planned')
    axs[4].plot(t_ctrl, hist_tau_applied[:, 1], color='#B45309', lw=1.8, label='$\\tau_r$ Applied')
    axs[4].set_xlabel('Time [s]', fontweight='bold')
    axs[4].set_ylabel('Forces', fontweight='bold')
    axs[4].set_title('Control Forces', fontsize=11, fontweight='bold', pad=6)
    axs[4].grid(True, ls=':', alpha=0.6)
    axs[4].legend(loc='upper right', ncol=2, frameon=True, facecolor='white', fontsize=8.5)

    axs[5].plot(t_sim, T_plan[:, 0], color=c_green_plan, ls='--', lw=1.8, label='$T_1$ Plan')
    axs[5].plot(t_ctrl, hist_T_applied[:, 0], color=c_green_real, lw=1.8, label='$T_1$ Applied')
    axs[5].plot(t_sim, T_plan[:, 1], color='#F43F5E', ls='--', lw=1.8, label='$T_2$ Plan')
    axs[5].plot(t_ctrl, hist_T_applied[:, 1], color=c_real, lw=1.8, label='$T_2$ Applied')
    axs[5].axhline(T_MAX, color='#9CA3AF', ls=':', lw=1.2, label=f'$T_{{max}}$ ({T_MAX:.1f} N)')
    axs[5].axhline(T_MIN, color='#9CA3AF', ls=':', lw=1.2, label=f'$T_{{min}}$ ({T_MIN:.1f} N)')
    axs[5].set_xlabel('Time [s]', fontweight='bold')
    axs[5].set_ylabel('Thrust [N]', fontweight='bold')
    axs[5].set_title('Thruster Allocation', fontsize=11, fontweight='bold', pad=6)
    axs[5].grid(True, ls=':', alpha=0.6)
    axs[5].legend(loc='upper right', ncol=3, frameon=True, facecolor='white', fontsize=8.0)

    plt.tight_layout()
    out_img_path = os.path.join(script_dir, 'c2_openloop_simulation_results.png')
    plt.savefig(out_img_path, dpi=300, bbox_inches='tight')
    plt.close()
    print(f"Open-loop simulation completed. Plot saved to: {out_img_path}")

if __name__ == '__main__':
    main()

File: case2_semilla_c/src/c2_simulate_openloop.c
/* Technique: Case 2 - Minimum Jerk QP Trajectory Planning +
 * 6-Parameter Pseudo-Flatness Reconstruction (m11 != m22)
 * C port of c2_simulate_openloop.py: plans with the algebraic (non-integrating)
 * psi solve, then feeds tau_act through an RK4 truth-model simulation to
 * measure open-loop tracking error. */

#define _POSIX_C_SOURCE 199309L
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#include "c2_usv_params.h"
#include "c2_min_jerk_qp.h"
#include "c2_flatness_reconstruct.h"

#ifndef M_PI
#define M_PI 3.14159265358979323846
#endif

#define N_WP 9

static double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000.0 + ts.tv_nsec / 1e6;
}

/* state = [x, y, psi, u, v, r] */
static void deriv(const double s[6], double Tu, double Tr, double out[6]) {
    double psi = s[2], u = s[3], v = s[4], r = s[5];
    out[0] = u * cos(psi) - v * sin(psi);
    out[1] = u * sin(psi) + v * cos(psi);
    out[2] = r;
    out[3] = (Tu + m22_p * v * r - Xu_p * u) / m11_p;
    out[4] = (-m11_p * u * r - Yv_p * v) / m22_p;
    out[5] = (Tr + (m11_p - m22_p) * u * v - Nr_p * r) / m33_p;
}

static void rk4_step(double s[6], double Tu, double Tr, double dt) {
    double k1[6], k2[6], k3[6], k4[6], tmp[6];
    deriv(s, Tu, Tr, k1);
    for (int i = 0; i < 6; i++) tmp[i] = s[i] + 0.5 * dt * k1[i];
    deriv(tmp, Tu, Tr, k2);
    for (int i = 0; i < 6; i++) tmp[i] = s[i] + 0.5 * dt * k2[i];
    deriv(tmp, Tu, Tr, k3);
    for (int i = 0; i < 6; i++) tmp[i] = s[i] + dt * k3[i];
    deriv(tmp, Tu, Tr, k4);
    for (int i = 0; i < 6; i++)
        s[i] += (dt / 6.0) * (k1[i] + 2 * k2[i] + 2 * k3[i] + k4[i]);
}

int main(void) {
    double waypoints_x[N_WP] = {0.0, 8.0, 14.0, 14.0, 20.0, 28.0, 32.0, 26.0, 20.0};
    double waypoints_y[N_WP] = {0.0, 0.0, 5.0, 13.0, 17.0, 17.0, 10.0, 4.0, 1.5};
    double base_times[N_WP]  = {0.0, 7.0, 14.0, 20.0, 27.0, 33.0, 40.0, 48.0, 55.0};
    double time_scale = 2.10;
    double times[N_WP];
    for (int i = 0; i < N_WP; i++) times[i] = base_times[i] * time_scale;

    double v0x = 0.1, v0y = 0.0;
    double dfx = waypoints_x[N_WP - 1] - waypoints_x[N_WP - 2];
    double dfy = waypoints_y[N_WP - 1] - waypoints_y[N_WP - 2];
    double dn = hypot(dfx, dfy);
    double vfx = 0.01 * dfx / dn, vfy = 0.01 * dfy / dn;

    double t_start_ms = now_ms();

    MinJerk2D planner;
    minjerk2d_solve(&planner, waypoints_x, waypoints_y, times, N_WP, v0x, v0y, vfx, vfy);

    double *t_sim, *px, *py, *vx, *vy, *ax, *ay, *jx, *jy;
    int N;
    minjerk2d_sample(&planner, DT_SIM, &t_sim, &px, &py, &vx, &vy, &ax, &ay, &jx, &jy, &N);

    FlatnessResult flat;
    reconstruct_flatness_h2(N, t_sim, px, py, vx, vy, ax, ay, &flat);

    double solve_time_ms = now_ms() - t_start_ms;

    /* --- open-loop RK4 truth-model rollout, driven by tau_act --- */
    double *real_x = malloc(sizeof(double) * (size_t)N);
    double *real_y = malloc(sizeof(double) * (size_t)N);
    double *real_psi = malloc(sizeof(double) * (size_t)N);
    double *real_u = malloc(sizeof(double) * (size_t)N);
    double *real_v = malloc(sizeof(double) * (size_t)N);
    double *real_r = malloc(sizeof(double) * (size_t)N);

    double state[6] = { flat.x[0], flat.y[0], flat.psi[0], flat.u[0], flat.v[0], flat.r[0] };
    real_x[0] = state[0]; real_y[0] = state[1]; real_psi[0] = state[2];
    real_u[0] = state[3]; real_v[0] = state[4]; real_r[0] = state[5];

    for (int i = 0; i < N - 1; i++) {
        double Tu = flat.tau_u_act[i], Tr = flat.tau_r_act[i];
        rk4_step(state, Tu, Tr, DT_SIM);
        real_x[i + 1] = state[0]; real_y[i + 1] = state[1]; real_psi[i + 1] = state[2];
        real_u[i + 1] = state[3]; real_v[i + 1] = state[4]; real_r[i + 1] = state[5];
    }

    /* --- tracking error metrics --- */
    double sum_pos2 = 0.0, max_pos = 0.0;
    double sum_psi2 = 0.0;
    double sum_u2 = 0.0, sum_v2 = 0.0, sum_r2 = 0.0;
    for (int i = 0; i < N; i++) {
        double dx = real_x[i] - flat.x[i], dy = real_y[i] - flat.y[i];
        double dpos = hypot(dx, dy);
        sum_pos2 += dpos * dpos;
        if (dpos > max_pos) max_pos = dpos;

        double dpsi = atan2(sin(real_psi[i] - flat.psi[i]), cos(real_psi[i] - flat.psi[i]));
        sum_psi2 += dpsi * dpsi;

        double du = real_u[i] - flat.u[i], dv = real_v[i] - flat.v[i], dr = real_r[i] - flat.r[i];
        sum_u2 += du * du; sum_v2 += dv * dv; sum_r2 += dr * dr;
    }
    double rmse_pos = sqrt(sum_pos2 / N);
    double rmse_psi_rad = sqrt(sum_psi2 / N);
    double rmse_psi_deg = rmse_psi_rad * 180.0 / M_PI;
    double rmse_u = sqrt(sum_u2 / N);
    double rmse_v = sqrt(sum_v2 / N);
    double rmse_r = sqrt(sum_r2 / N);

    printf("\n==================================================\n");
    printf("CASE 2 OPEN-LOOP TRAJECTORY METRICS (C port):\n");
    printf("  - Trajectory Solver Time (QP): %.4f ms\n", solve_time_ms);
    printf("  - Position RMSE:               %.4f m\n", rmse_pos);
    printf("  - Max Position Error:          %.4f m\n", max_pos);
    printf("  - Heading RMSE:                %.4f rad (%.2f deg)\n", rmse_psi_rad, rmse_psi_deg);
    printf("  - Surge Velocity (u) RMSE:     %.4f m/s\n", rmse_u);
    printf("  - Sway Velocity (v) RMSE:      %.4f m/s\n", rmse_v);
    printf("  - Yaw Rate (r) RMSE:           %.4f rad/s\n", rmse_r);
    printf("==================================================\n\n");

    /* --- c2_openloop_results.csv (planned reference) --- */
    FILE *fr = fopen("c2_openloop_results.csv", "w");
    if (fr) {
        fprintf(fr, "t,x,y,psi,u,v,r,tau_u,tau_r,T1,T2\n");
        for (int i = 0; i < N; i++)
            fprintf(fr, "%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n",
                    t_sim[i], flat.x[i], flat.y[i], flat.psi[i],
                    flat.u[i], flat.v[i], flat.r[i], flat.tau_u[i], flat.tau_r[i],
                    flat.T1_dem[i], flat.T2_dem[i]);
        fclose(fr);
    }

    /* --- c2_openloop_applied.csv (applied controls + resulting real state) --- */
    FILE *fa = fopen("c2_openloop_applied.csv", "w");
    if (fa) {
        fprintf(fa, "t,tau_u_applied,tau_r_applied,T1_applied,T2_applied,real_x,real_y,real_psi,real_u,real_v,real_r\n");
        for (int i = 0; i < N; i++) {
            double tu = (i < N - 1) ? flat.tau_u_act[i] : flat.tau_u_act[N - 2];
            double tr = (i < N - 1) ? flat.tau_r_act[i] : flat.tau_r_act[N - 2];
            double T1a = (i < N - 1) ? flat.T1_act[i] : flat.T1_act[N - 2];
            double T2a = (i < N - 1) ? flat.T2_act[i] : flat.T2_act[N - 2];
            fprintf(fa, "%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f,%.6f\n",
                    t_sim[i], tu, tr, T1a, T2a,
                    real_x[i], real_y[i], real_psi[i], real_u[i], real_v[i], real_r[i]);
        }
        fclose(fa);
    }

    /* --- openloop_metrics.json --- */
    FILE *fm = fopen("openloop_metrics.json", "w");
    if (fm) {
        fprintf(fm,
            "{\n"
            "    \"case\": \"Case 2 Semilla (C)\",\n"
            "    \"solver_type\": \"QP (6-Param Pseudo-Flatness) - C port\",\n"
            "    \"trajectory_solver_time_ms\": %.6f,\n"
            "    \"tracking_error\": {\n"
            "        \"rmse_position_m\": %.6f,\n"
            "        \"max_position_error_m\": %.6f,\n"
            "        \"rmse_heading_rad\": %.6f,\n"
            "        \"rmse_heading_deg\": %.6f,\n"
            "        \"rmse_surge_u_mps\": %.6f,\n"
            "        \"rmse_sway_v_mps\": %.6f,\n"
            "        \"rmse_yaw_rate_r_radps\": %.6f\n"
            "    }\n"
            "}\n",
            solve_time_ms, rmse_pos, max_pos, rmse_psi_rad, rmse_psi_deg, rmse_u, rmse_v, rmse_r);
        fclose(fm);
    }

    printf("c2_openloop_results.csv, c2_openloop_applied.csv, openloop_metrics.json written.\n");
    printf("Run 'python3 plot_results.py --openloop' to regenerate the diagnostic PNG.\n");

    free(real_x); free(real_y); free(real_psi); free(real_u); free(real_v); free(real_r);
    flatness_result_free(&flat);
    free(t_sim); free(px); free(py); free(vx); free(vy); free(ax); free(ay); free(jx); free(jy);
    minjerk2d_free(&planner);
    return 0;
}
